MPE Continuous Action support

Rohan138 opened the pull request:
I've added support for continuous actions in MPE as discussed in #249 through the continuous_actions=False argument in the environment config.

I've tested my changes on all environments using RLLib MADDPG, and run ./release_test.sh. The latter needed some updates as well.

All environments are working except simple_world_comm_v2, which seems to have a supersuit-related bug.

Rohan138 committed:
Added continuous action support to MPE

Rohan138 committed:
Update setup.py

Rohan138 committed:
Cleanup and docs for MPE continuous actions

Rohan138 committed:
Merge branch 'master' of https://github.com/Rohan138/PettingZoo

Rohan138 committed:
Fix pip bug

Rohan138 committed:
Passing flake8 tests

Rohan138 committed:
Ran release_test.sh

jkterry1 commented:
Hey I haven't been tracking your conversation on discord, but you understand that our MPE environments are rather substantially fixed upband can't be directly compared to the ones in the original repo right?

jkterry1 commented:
And does Ben have the needed information to look into the "supersuit related bug"?

Rohan138 committed:
Fixed docs for simple_world_comm

Rohan138 commented:
Yes, I understand-however, I think having the option of using continuous action spaces makes it much easier to use PettingZoo MPE with most papers using MPE, just because they've been using it as well. I've still kept the default actions to discrete, so the original behavior isn't affected.
For (example)[https://gist.github.com/Rohan138/1dd2f0063b8a6150c8fa75544af28e3c]

Rohan138 committed:
Test continuous_actions=True in all_parameter_combs.py

benblack769 commented:
So I enabled CI tests for your PR, and it looks like `simple_world_comm_v3` is failing due to the observation space issue. You can run the respective test with 

```
pytest ./test/pytest_runner.py::test_module[mpe/simple_world_comm_v2-pettingzoo.mpe.simple_world_comm_v2]
```

Let me know if you have any questions.

benblack769 commented:
Note that supersuit is not in those tests, so it is not a supersuit issue.

benblack769 commented:
Ok, so I am reviewing the PR, and it generally looks good, really good job figuring out the tests and documentation.

One issue is that the continuous action space for environments with both communication and movement (like simple_world_comm_v2) is not what it should be (this might be what is causing the test to fail).

In particular, if you have 5 movement options and 4 communication options, then in a continuous space, the action should be of size 9 (the actions are concatenated), and in a discrete space, it should be of size 20 (the actions are a cross product of the two subspaces). The continuous space should break the action down into the movement and communication components differently in the continuous and discrete cases here: https://github.com/PettingZoo-Team/PettingZoo/blob/master/pettingzoo/mpe/_mpe_utils/simple_env.py#L94

The documentation should reflect this.

Thanks again for working on this!

Rohan138 committed:
Fixed docs for MPE

Rohan138 commented:
Hi, so I'd actually implemented that correctly, but I hadn't added that to the docs, which I just fixed. simple_reference, which also has an agent that can both move and communicate, which works fine.
I'm trying to track down the issue, but I don't see how my changes are affecting the observation space, since all my code changes have been in the _set_action function in [simple_env.py](https://github.com/PettingZoo-Team/PettingZoo/blob/master/pettingzoo/mpe/_mpe_utils/simple_env.py#L94)?

Rohan138 committed:
Fixed continuous action spaces

Rohan138 committed:
Fixed action spaces

Rohan138 committed:
Fixed simple_world_comm_v2 bug

Rohan138 committed:
Removed test_output.txt

benblack769 commented:
Ok, so the action space and action handling looks good now. I don't like the way that the rendering misrepresents the communication (it renders the argmax of the actions), but that is inherited from the original MPE, and I don't know if its worth fixing for now.

So I am happy with this PR. @jkterry1 Do you want to look over the documentation? I think its fine, but I have low standards, as you know.

jkterry1 commented:
"I don't like the way that the rendering misrepresents the communication (it renders the argmax of the actions), but that is inherited from the original MPE, and I don't know if its worth fixing for now."

@Rohan138 would you be willing to work on this?

jkterry1 commented:
@benblack769 the docs look fine..? what body am I supposed to be looking for under the rug here

benblack769 commented:
@jkterry1  I don't see any dead bodies either, I'm just checking that you don't.

Rohan138 committed:
Added better rendering for continuous MPE

Rohan138 committed:
Added mpe_maddpg.py to tutorials

Rohan138 commented:
I added something-just printing out the entire comm, rounded to 2dp - that's the best thing I could think of, although I agree that the argmax is unsatisfying. Perhaps a different PR?
I also added the mpe_maddpg script to tutorials.
Feel free to reject either of these two commits.

Rohan138 committed:
Fixed example

benblack769 commented:
Can you check the "Allow edits from maintainers" box? There are a couple minor things I want to change.
https://github.com/bisq-network/style/issues/4

Rohan138 commented:
It's already checked, are you unable to edit?

jkterry1 commented:
Ben that feature really doesn't work reliably on GitHub. It's a recurring
problem with my PRs to external repos too that I've never seen a solution
for.

On Sun, Jul 11, 2021 at 7:53 PM Rohan138 ***@***.***> wrote:

> It's already checked, are you unable to edit?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/PettingZoo-Team/PettingZoo/pull/419#issuecomment-877930115>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEUF33D2QI2NK4G4WR3UPWTTXJKJTANCNFSM5AAAFO7Q>
> .
>
-- 
Thank you for your time,
Justin Terry

Rohan138 commented:
Hi, any updates on this?

Rohan138 committed:
Merge branch 'PettingZoo-Team:master' into master

jkterry1 committed:
Delete mpe_maddpg.py

jkterry1 commented:
I'll try to meet with Ben today and deal with this

benblack769 committed:
fixed rendering

benblack769 committed:
Merge branch 'master' of https://github.com/Rohan138/PettingZoo

benblack769 commented:
Sorry for ignoring this. I just fiddled around a bit with the rendering, no real complains.

benblack769 commented:
So I am happy with this being merged

benblack769 commented:
@jkterry1
